File: mosaic-tty/src/jvmMain/java/com/jakewharton/mosaic/tty/Jni.java
package com.jakewharton.mosaic.tty;

final class Jni {
	static {
		NativeLibrary.ensureLoaded();
	}

	static native long streamsInit();

	static native boolean streamsInputIsTty(long streamsPtr);

	static native boolean streamsOutputIsTty(long streamsPtr);

	static native boolean streamsErrorIsTty(long streamsPtr);

	static native int streamsReadInput(
		long streamsPtr,
		byte[] buffer,
		int offset,
		int count
	);

	static native int streamsReadInputWithTimeout(
		long streamsPtr,
		byte[] buffer,
		int offset,
		int count,
		int timeout
	);

	static native void streamsInterruptInputRead(long streamsPtr);

	static native int streamsWriteOutput(
		long streamsPtr,
		byte[] buffer,
		int offset,
		int count
	);

	static native int streamsWriteError(
		long streamsPtr,
		byte[] buffer,
		int offset,
		int count
	);

	static native void streamsFree(long streamsPtr);

	static native long ttyCallbackInit(Tty.Callback callback);

	static native void ttyCallbackFree(long callbackPtr);

	static native long ttyInit();

	static native void ttySetCallback(long ttyPtr, long callbackPtr);

	static native int ttyRead(
		long ttyPtr,
		byte[] buffer,
		int offset,
		int count
	);

	static native int ttyReadWithTimeout(
		long ttyPtr,
		byte[] buffer,
		int offset,
		int count,
		int timeoutMillis
	);

	static native void ttyInterruptRead(long ttyPtr);

	static native int ttyWrite(
		long ttyPtr,
		byte[] buffer,
		int offset,
		int count
	);

	static native void ttyEnableRawMode(long ttyPtr);

	static native void ttyEnableWindowResizeEvents(long ttyPtr);

	/**
	 * @return Array of `[columns, rows, width, height]`. Using an array saves us from having to
	 * pass a complex object across the JNI boundary.
	 */
	static native int[] ttyCurrentSize(long ttyPtr);

	static native void ttyReset(long ttyPtr);

	static native void ttyFree(long ttyPtr);

	static native long testInit(boolean stdinIsTty, boolean stdoutIsTty, boolean stderrIsTty);

	static native long testGetStreams(long testTtyPtr);

	static native long testGetTty(long testTtyPtr);

	static native int testWriteTty(long testTtyPtr, byte[] buffer, int offset, int count);

	static native int testReadTty(long testTtyPtr, byte[] buffer, int offset, int count);

	static native int testReadTtyWithTimeout(long testTtyPtr, byte[] buffer, int offset, int count, int timeoutMillis);

	static native void testInterruptTtyRead(long testTtyPtr);

	static native int testWriteInput(long testTtyPtr, byte[] buffer, int offset, int count);

	static native int testReadOutput(long testTtyPtr, byte[] buffer, int offset, int count);

	static native int testReadOutputWithTimeout(long testTtyPtr, byte[] buffer, int offset, int count, int timeoutMillis);

	static native void testInterruptOutputRead(long testTtyPtr);

	static native int testReadError(long testTtyPtr, byte[] buffer, int offset, int count);

	static native int testReadErrorWithTimeout(long testTtyPtr, byte[] buffer, int offset, int count, int timeoutMillis);

	static native void testInterruptErrorRead(long testTtyPtr);

	static native void testResize(
		long testTtyPtr,
		int columns,
		int rows,
		int width,
		int height
	);

	static native void testSendFocusEvent(long testTtyPtr, boolean focused);

	static native void testSendKeyEvent(long testTtyPtr);

	static native void testSendMouseEvent(long testTtyPtr);

	static native void testFree(long testTtyPtr);

	private Jni() {}
}

File: mosaic-tty/src/jvmMain/kotlin/com/jakewharton/mosaic/tty/Jni.kt
package com.jakewharton.mosaic.tty

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption.REPLACE_EXISTING
import java.util.Locale.US

// TODO @JvmSynthetic https://youtrack.jetbrains.com/issue/KT-24981
internal object Jni {
	init {
		loadNativeLibrary("mosaic")
	}

	@JvmStatic
	external fun ttyCallbackInit(callback: Tty.Callback): Long

	@JvmStatic
	external fun ttyCallbackFree(callbackPtr: Long)

	@JvmStatic
	external fun ttyInit(callbackPtr: Long): Long

	@JvmStatic
	external fun ttyRead(
		ttyPtr: Long,
		buffer: ByteArray,
		offset: Int,
		count: Int,
	): Int

	@JvmStatic
	external fun ttyReadWithTimeout(
		ttyPtr: Long,
		buffer: ByteArray,
		offset: Int,
		count: Int,
		timeoutMillis: Int,
	): Int

	@JvmStatic
	external fun ttyInterrupt(ttyPtr: Long)

	@JvmStatic
	external fun ttyEnableRawMode(ttyPtr: Long)

	@JvmStatic
	external fun ttyEnableWindowResizeEvents(ttyPtr: Long)

	/**
	 * @return Array of `[columns, rows, width, height]`. Using an array saves us from having to
	 * pass a complex object across the JNI boundary.
	 */
	@JvmStatic
	external fun ttyCurrentSize(ttyPtr: Long): IntArray

	@JvmStatic
	external fun ttyFree(ttyPtr: Long)

	@JvmStatic
	external fun testTtyInit(callbackPtr: Long): Long

	@JvmStatic
	external fun testTtyGetTty(testTtyPtr: Long): Long

	@JvmStatic
	external fun testTtyWrite(testTtyPtr: Long, buffer: ByteArray)

	@JvmStatic
	external fun testTtyFocusEvent(testTtyPtr: Long, focused: Boolean)

	@JvmStatic
	external fun testTtyKeyEvent(testTtyPtr: Long)

	@JvmStatic
	external fun testTtyMouseEvent(testTtyPtr: Long)

	@JvmStatic
	external fun testTtyResizeEvent(
		testTtyPtr: Long,
		columns: Int,
		rows: Int,
		width: Int,
		height: Int,
	)

	@JvmStatic
	external fun testTtyFree(testTtyPtr: Long)

	@Suppress(
		// Only loading from our own JAR contents.
		"UnsafeDynamicallyLoadedCode",
		// Preserving copy/paste!
		"SameParameterValue",
	)
	private fun loadNativeLibrary(name: String) {
		val osName = System.getProperty("os.name").lowercase(US)
		val osArch = System.getProperty("os.arch").lowercase(US)
		val nativeLibraryJarPath = "/jni/$osArch/" + when {
			"linux" in osName -> "lib$name.so"
			"mac" in osName -> "lib$name.dylib"
			"windows" in osName -> "$name.dll"
			else -> throw IllegalStateException("Unsupported OS: $osName $osArch")
		}
		val nativeLibraryUrl = Jni::class.java.getResource(nativeLibraryJarPath)
			?: throw IllegalStateException("Unable to read $nativeLibraryJarPath from JAR")
		val nativeLibraryFile: Path
		try {
			nativeLibraryFile = Files.createTempFile(name, null)

			// File-based deleteOnExit() uses a special internal shutdown hook that always runs last.
			nativeLibraryFile.toFile().deleteOnExit()
			nativeLibraryUrl.openStream().use { nativeLibrary ->
				Files.copy(nativeLibrary, nativeLibraryFile, REPLACE_EXISTING)
			}
		} catch (e: IOException) {
			throw RuntimeException("Unable to extract native library from JAR", e)
		}
		System.load(nativeLibraryFile.toAbsolutePath().toString())
	}
}
